### PR TITLE
Spark to 2.3.0 and update other dependencies

### PR DESCRIPTION
**Pull request by jklukas**
This brings back in the Spark version update merged in #426 and subsequently backed out.

It also updates several other libraries to more recent (most recent where possible) versions.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -18,28 +20,29 @@ lazy val root = (project in file(".")).
     scalaModuleInfo := scalaModuleInfo.value.map(_.withOverrideScalaVersion(true)),
     libraryDependencies += "com.mozilla.telemetry" %% "moztelemetry" % "1.1-SNAPSHOT",
     libraryDependencies += "com.mozilla.telemetry" %% "spark-hyperloglog" % "2.0.0-SNAPSHOT",
-    libraryDependencies += "org.apache.avro" % "avro" % "1.7.7",
-    libraryDependencies += "org.apache.parquet" % "parquet-avro" % "1.7.0",
-    libraryDependencies += "org.scalatest" %% "scalatest" % "3.0.4",
```
Comment (jklukas) [inline build.sbt]: Moved all test dependencies to the bottom of the list and ensured they're only included in Test scope.

File: build.sbt [hunk 2]
```diff
@@ -18,28 +20,29 @@ lazy val root = (project in file(".")).
...
+    libraryDependencies += "org.apache.avro" % "avro" % "1.8.2",
+    libraryDependencies += "org.apache.parquet" % "parquet-avro" % "1.8.3",
     libraryDependencies += "net.sandrogrzicic" %% "scalabuff-runtime" % "1.4.0",
-    libraryDependencies += "com.holdenkarau" %% "spark-testing-base" % "2.0.0_0.4.7" % "test",
-    libraryDependencies += "org.xerial.snappy" % "snappy-java" % "1.1.2",
-    libraryDependencies += "org.json4s" %% "json4s-jackson" % "3.2.10",
-    libraryDependencies += "joda-time" % "joda-time" % "2.9.2",
-    libraryDependencies += "org.apache.hadoop" % "hadoop-client" % "2.7.1" excludeAll(ExclusionRule(organization = "javax.servlet")),
-    libraryDependencies += "org.apache.hadoop" % "hadoop-aws" % "2.7.1" excludeAll(ExclusionRule(organization = "javax.servlet")),
-    libraryDependencies += "com.github.nscala-time" %% "nscala-time" % "2.10.0",
```
Comment (jklukas) [inline build.sbt]: It appears we don't use nscala-time anymore, so removed it.

File: build.sbt [hunk 3]
```diff
@@ -18,28 +20,29 @@ lazy val root = (project in file(".")).
...
+    libraryDependencies += "org.apache.hadoop" % "hadoop-aws" % hadoopVersion excludeAll(ExclusionRule(organization = "javax.servlet")),
     libraryDependencies += "org.rogach" %% "scallop" % "3.1.2",
     libraryDependencies += "org.apache.spark" %% "spark-core" % sparkVersion,
     libraryDependencies += "org.apache.spark" %% "spark-sql" % sparkVersion,
     libraryDependencies += "org.apache.spark" %% "spark-mllib" % sparkVersion,
     libraryDependencies += "org.apache.spark" %% "spark-hive" % sparkVersion,
-    libraryDependencies += "org.scalaj" %% "scalaj-http" % "2.3.0",
+    libraryDependencies += "org.scalaj" %% "scalaj-http" % "2.4.0",
     libraryDependencies += "org.yaml" % "snakeyaml" % "1.21",
-    libraryDependencies += "com.google.protobuf" % "protobuf-java" % "2.5.0",
```
Comment (jklukas) [inline build.sbt]: We don't directly use this dependency, so removing it from the list here and instead just relying on the override.

File: run-sbt.sh [hunk 4]
```diff
@@ -12,15 +12,18 @@ if [ ! -f /.dockerenv ]; then
     exit $?
 fi
 
+# Set options that sbt will pass to the JVM
+export SBT_OPTS="-Xss2M -Djava.security.policy=java.policy"
```
Comment (jklukas) [inline run-sbt.sh]: The security policy is necessary due to a change in derby as discussed in #426

**Comment by jklukas** [comment]
Ran main_summary on nightly data using the master branch and this branch. The DatasetComparator confirmed that the output datasets were the same.

**Comment by jklukas** [comment]
Well, I tested this and then changed just a _few little things_. Tests are now failing, so this will likely need to wait for next week.

File: run-sbt.sh (hunk 4 above)
Comment (fbertsch) [reply run-sbt.sh]: It might be good to put a link here for reference.

File: build.sbt [hunk 5]
```diff
@@ -16,30 +18,38 @@ lazy val root = (project in file(".")).
     // Hack to get releases to not fail to upload when the same jar name already exists. Later we will need auto versioning
     isSnapshot := true,
     scalaModuleInfo := scalaModuleInfo.value.map(_.withOverrideScalaVersion(true)),
+
+    // Snapshot dependencies; keep in mind that changes to the master branch of these projects will get pulled in
+    // automatically on next build and could cause unexpected problems while working on unrelated changes.
     libraryDependencies += "com.mozilla.telemetry" %% "moztelemetry" % "1.1-SNAPSHOT",
     libraryDependencies += "com.mozilla.telemetry" %% "spark-hyperloglog" % "2.0.0-SNAPSHOT",
-    libraryDependencies += "org.apache.avro" % "avro" % "1.7.7",
-    libraryDependencies += "org.apache.parquet" % "parquet-avro" % "1.7.0",
```
Comment (jklukas) [inline build.sbt]: These dependencies have moved further down the list and have updated versions

File: build.sbt [hunk 6]
```diff
@@ -16,30 +18,38 @@ lazy val root = (project in file(".")).
...
-    libraryDependencies += "org.scalatest" %% "scalatest" % "3.0.4",
-    libraryDependencies += "net.sandrogrzicic" %% "scalabuff-runtime" % "1.4.0",
-    libraryDependencies += "com.holdenkarau" %% "spark-testing-base" % "2.0.0_0.4.7" % "test",
-    libraryDependencies += "org.xerial.snappy" % "snappy-java" % "1.1.2",
-    libraryDependencies += "org.json4s" %% "json4s-jackson" % "3.2.10",
```
Comment (jklukas) [inline build.sbt]: Removed the json4s-jackson so that we get the version spark-core wants transitively.

File: build.sbt [hunk 7]
```diff
@@ -53,7 +63,7 @@ test in assembly := {}
 
 assemblyMergeStrategy in assembly := {
   case PathList("META-INF", xs @ _*) => MergeStrategy.discard
-  case x => MergeStrategy.first
+  case _ => MergeStrategy.first
```
Comment (jklukas) [inline build.sbt]: Unrelated style fixup. Using underscore to indicate that we don't care about the value in this case.

File: run-sbt.sh (hunk 4 above)
Comment (jklukas) [reply run-sbt.sh]: There's a link in the java.policy file itself, and explanation of why it exists. Does that seem sufficient?

**Comment by jklukas** [comment]
This is passing again with `java.policy` taken out, so ready for review again, @fbertsch

**Pull request merged**